File: 2256.minimum-average-difference.cs
/*
 * @lc app=leetcode id=2256 lang=csharp
 *
 * [2256] Minimum Average Difference
 */

// @lc code=start
public class Solution 
{
    public int MinimumAverageDifference(int[] nums) 
    {
        int n = nums.Length;
        long[] prefix = new long[n + 1];
        prefix[0] = 0;
        for (int i = 0; i < n; i++)
        {
            prefix[i + 1] = prefix[i] + nums[i];
        }

        long minsofar = prefix[prefix.Length - 1] / n;
        int result = n - 1;

        for (int i = 0; i < n - 1; i++)
        {
            long current = Math.Abs(prefix[i+1] / (i+1) - (prefix[prefix.Length-1] - prefix[i+1]) / (n - i - 1));
            if (current < minsofar)
            {
                minsofar = current;
                result = i;
            }
            else if (current == minsofar && i < result)
            {
                result = i;
            }
        }
        return result;        
    }
}
// @lc code=end

File: 2256.minimum-average-difference.py
#
# @lc app=leetcode id=2256 lang=python3
#
# [2256] Minimum Average Difference
#

# @lc code=start
class Solution:
    # Prefix sum would handle the problem nicely
    def minimumAverageDifference(self, nums: List[int]) -> int:
        n = len(nums)
        prefix = [0]
        for num in nums:
            prefix.append(prefix[-1] + num)
        
        minsofar = prefix[-1] // n
        result = n - 1
        for i in range(n-1):
            current = abs(prefix[i+1] // (i + 1) - (prefix[-1] - prefix[i+1]) // (n - i - 1))
            if current < minsofar:
                minsofar = current
                result = i
            elif current == minsofar and i < result:
                result = i
        return result
                        
# @lc code=end

